tensorboard/java/org/tensorflow/tensorboard/vulcanize/Vulcanize.java

```diff
@@ -778,6 +778,55 @@ private static String extractAndTransformJavaScript(Document document, Webpath j
     return scriptContent;
   }
 
+  private static void cloneChildrenWithoutWhitespace(Element src, Element dest) {
+    List<Node> toMove = new ArrayList<Node>();
+    for (Node node : src.childNodes()) {
+      if (node instanceof TextNode && ((TextNode) node).isBlank()) {
+        continue;
+      }
+      toMove.add(node);
+    }
+    for (int i = 0; i < toMove.size(); i++) {
+      Node node = toMove.get(i);
+      dest.appendChild(node.clone());
+    }
+  }
+
+  // Refer to https://github.com/tensorflow/tensorboard/issues/3557.
+  private static Document getFlattenedDocument(Document document) {
+    Document flatDoc = new Document("/");
+    flatDoc.normalise();
+    Element rootDocumentHead = flatDoc.head();
+    Element rootDocumentBody = flatDoc.body();
+
+    Node currentNode = document;
+    while (currentNode != null) {
+      // Do not clone the element if it is a `document` inside `<head>`.
+      // We want to traverse further and get all elements from `<head>` and `<body>`
+      if (currentNode.parentNode() != null && currentNode.parentNode().nodeName().equals("head")
+          && !(currentNode instanceof Document)) {
+        rootDocumentHead.appendChild(currentNode.clone());
+      }
+
+      if (currentNode.nodeName().equals("body")) {
+        cloneChildrenWithoutWhitespace((Element) currentNode, rootDocumentBody);
+      }
+
+      if (currentNode.childNodeSize() > 0) {
+        currentNode = currentNode.childNode(0);
+      } else {
+        while (currentNode != null && currentNode.nextSibling() == null) {
+          currentNode = currentNode.parentNode();
```

Comment:
Is this the only place where currentNode might be set to null?  That would be worth mentioning to help understand when we can exit the top-level loop.